File: ruby/src/path-with-maximum-probability.rb
# @param {Integer} n
# @param {Integer[][]} edges
# @param {Float[]} succ_prob
# @param {Integer} start_node
# @param {Integer} end_node
# @return {Float}
def max_probability(n, edges, succ_prob, start_node, end_node)
  graph = Array.new(n) { [] }
  (0...edges.size).each do |i|
    prob = succ_prob[i]
    from, to = edges[i]
    graph[from].append([to, prob])
    graph[to].append([from, prob])
  end
  reachable = Containers::MaxHeap.new
  reachable.push([1.0, start_node])
  reached = Set.new
  while reachable.size > 0
    prob, node = reachable.pop
    return prob if node == end_node
    next if reached.include?(node)
    reached.add(node)

    graph[node].each do |adj, pr|
      reachable.push([prob * pr, adj])
    end
  end
  0.0
end

File: dart/path-with-maximum-probability.dart
import 'package:collection/collection.dart';

class Solution {
  double maxProbability(int n, List<List<int>> edges, List<double> succProb, int startNode, int endNode) {
    var graph = List<List<(int, double)>>.generate(n, (_) => [], growable: false);
    for (int i = 0; i < edges.length; i++) {
      var from = edges[i][0];
      var to = edges[i][1];
      var prob = succProb[i];
      graph[from].add((to, prob));
      graph[to].add((from, prob));
    }
    return dijkstra(graph, startNode, endNode);
  }

  double dijkstra(List<List<(int, double)>> graph, int from, int to) {
    var queue = PriorityQueue<(int, double)>((a, b) => b.$2.compareTo(a.$2));
    queue.add((from, 1.0));
    var visited = <int>{};
    while (!queue.isEmpty) {
      var (node, probability) = queue.removeFirst();
      if (node == to) {
        return probability;
      }
      if (visited.contains(node)) { continue; }

      visited.add(node);
      graph[node].forEach(((int, double) edge) {
        queue.add((edge.$1, probability * edge.$2));
      });
    }
    return 0.0;
  }
}